pynav/lib/imu.py

```diff
@@ -0,0 +1,477 @@
+
+    @property
+    def bias_accel(self) -> np.ndarray:
+        return self.value[2].value
+
+    @bias_accel.setter
+    def bias_accel(self, accel_bias):
+        self.value[2].value = accel_bias.ravel()
+
+    @property
+    def nav_state(self) -> np.ndarray:
+        return self.value[0].value
+
+    @property
+    def pose(self) -> np.ndarray:
+        return self.value[0].pose
+
+    @pose.setter
+    def pose(self, pose):
+        self.value[0].pose = pose
+
+    def plus(self, dx: np.ndarray, new_stamp: float = None):
+        """
+        Updates the value of each of the IMU state, given a perturbation dx.
+        """
+        if dx.shape[0] != 15:
+            raise ValueError("Perturbation must be dimension 15!")
+
+        for i, s in enumerate(self._slices):
+            sub_dx = dx[s]
+            self.value[i].plus(sub_dx)
+
+        return self.copy()
+
+    def minus(self, x: "IMUState") -> np.ndarray:
+        dx = []
+        for i, v in enumerate(x.value):
+            dx.append(self.value[i].minus(x.value[i]).reshape((-1, 1)))
+
+        return np.vstack(dx)
+
+    def copy(self):
+        """
+        Returns a new composite state object where the state values have also
+        been copied.
+        """
+        return IMUState(
+            self.nav_state.copy(),
+            self.bias_gyro.copy(),
+            self.bias_accel.copy(),
+            self.stamp,
+            self.state_id,
+            self.direction,
+        )
+
+    def jacobian_from_blocks(
+        self,
+        attitude: np.ndarray = None,
+        position: np.ndarray = None,
+        velocity: np.ndarray = None,
+        bias_gyro: np.ndarray = None,
+        bias_accel: np.ndarray = None,
+    ):
+        for jac in [attitude, position, velocity, bias_gyro, bias_accel]:
+            if jac is not None:
+                dim = jac.shape[0]
+                break
+
+        nav_jacobian = self.value[0].jacobian_from_blocks(
+            attitude=attitude, position=position, velocity=velocity
+        )
+        if bias_gyro is None:
+            bias_gyro = np.zeros((dim, 3))
+        if bias_accel is None:
+            bias_accel = np.zeros((dim, 3))
+
+        return np.hstack([nav_jacobian, bias_gyro, bias_accel])
+
+
+class IMUKinematics(ProcessModel):
+    def __init__(self, Q: np.ndarray, g_a=[0, 0, -9.80665]):
+        """Instantiate the IMUKinematic model.
+
+        Parameters
+        ----------
+        Q : np.ndarray
+            Discrete-time noise matrix.
+        g_a : np.ndarray
+            Gravity vector resolved in the inertial frame.
+            If None, default value is set to [0; 0; -9.81].
+        """
+        self._Q = Q
+
+        self._gravity = np.array(g_a).ravel()
+
+    def evaluate(self, x: IMUState, u: IMU, dt: float) -> IMUState:
+        """Propagates an IMU state forward one timestep from an IMU measurement.
+
+        The continuous-time IMU equations are discretized using the assumption
```

Comment:
Here this should be changed to: "The continuous-time IMU equations are discretized using the assumption that the IMU measurements are constant between two timesteps".